File: packages/android/app/src/main/java/com/byoky/app/ui/screens/WalletStatsSheet.kt
package com.byoky.app.ui.screens

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.byoky.app.data.Credential
import com.byoky.app.data.Gift
import com.byoky.app.data.GiftedCredential
import com.byoky.app.data.Provider
import com.byoky.app.data.RequestLog
import com.byoky.app.data.formatTokens
import com.byoky.app.data.giftBudgetPercent
import com.byoky.app.data.giftBudgetRemaining
import com.byoky.app.ui.theme.*

sealed class WalletStatsTarget {
    data class Credential(val credentialId: String) : WalletStatsTarget()
    data class Gift(val giftedCredentialId: String) : WalletStatsTarget()
}

private enum class StatsRange(val label: String, val millis: Long?) {
    DAY("24h", 86_400_000L),
    WEEK("7d", 604_800_000L),
    MONTH("30d", 2_592_000_000L),
    ALL("All", null),
}

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun WalletStatsSheet(
    target: WalletStatsTarget,
    credentials: List<Credential>,
    gifts: List<Gift>,
    giftedCredentials: List<GiftedCredential>,
    requestLogs: List<RequestLog>,
    giftPeerOnline: Map<String, Boolean>,
    onDismiss: () -> Unit,
) {
    ModalBottomSheet(
        onDismissRequest = onDismiss,
        containerColor = BgRaised,
    ) {
        when (target) {
            is WalletStatsTarget.Credential -> {
                val cred = credentials.firstOrNull { it.id == target.credentialId }
                if (cred == null) MissingStats()
                else CredentialStatsContent(
                    credential = cred,
                    gifts = gifts.filter { it.credentialId == cred.id },
                    requestLogs = requestLogs,
                )
            }
            is WalletStatsTarget.Gift -> {
                val gc = giftedCredentials.firstOrNull { it.id == target.giftedCredentialId }
                if (gc == null) MissingStats()
                else GiftStatsContent(
                    credential = gc,
                    online = giftPeerOnline[gc.giftId],
                )
            }
        }
    }
}

@Composable
private fun MissingStats() {
    Column(
        modifier = Modifier
            .fillMaxWidth()
            .padding(32.dp),
        horizontalAlignment = Alignment.CenterHorizontally,
    ) {
        Text("No data", color = TextSecondary)
    }
}

@Composable
private fun CredentialStatsContent(
    credential: Credential,
    gifts: List<Gift>,
    requestLogs: List<RequestLog>,
) {
    val provider = Provider.find(credential.providerId)
    var range by remember { mutableStateOf(StatsRange.WEEK) }

    val filtered = remember(requestLogs, range, credential.providerId) {
        val now = System.currentTimeMillis()
        val cutoff = range.millis?.let { now - it }
        requestLogs.filter {
            it.providerId == credential.providerId &&
                it.statusCode < 400 &&
                (cutoff == null || it.timestamp > cutoff)
        }
    }

    val totalInput = filtered.sumOf { it.inputTokens ?: 0 }
    val totalOutput = filtered.sumOf { it.outputTokens ?: 0 }

    data class ModelRow(val model: String, val requests: Int, val total: Int)
    data class AppRow(val origin: String, val requests: Int, val total: Int)

    val byModel = remember(filtered) {
        val map = mutableMapOf<String, IntArray>() // [requests, tokens]
        for (e in filtered) {
            val model = e.model ?: continue
            val arr = map.getOrPut(model) { IntArray(2) }
            arr[0] += 1
            arr[1] += (e.inputTokens ?: 0) + (e.outputTokens ?: 0)
        }
        map.map { ModelRow(it.key, it.value[0], it.value[1]) }.sortedByDescending { it.total }
    }

    val byApp = remember(filtered) {
        val map = mutableMapOf<String, IntArray>() // [requests, tokens]
        for (e in filtered) {
            val arr = map.getOrPut(e.appOrigin) { IntArray(2) }
            arr[0] += 1
            arr[1] += (e.inputTokens ?: 0) + (e.outputTokens ?: 0)
        }
        map.map { AppRow(it.key, it.value[0], it.value[1]) }.sortedByDescending { it.total }
    }

    Column(
        modifier = Modifier
            .fillMaxWidth()
            .verticalScroll(rememberScrollState())
            .padding(horizontal = 24.dp)
            .padding(bottom = 32.dp),
        verticalArrangement = Arrangement.spacedBy(14.dp),
    ) {
        Column {
            Text(
                credential.label,
                fontSize = 20.sp,
                fontWeight = FontWeight.Bold,
                color = TextPrimary,
            )
            Text(
                "${provider?.name ?: credential.providerId} · shared across credentials of this provider",
                fontSize = 12.sp,
                color = TextSecondary,
            )
        }

        RangePicker(selected = range, onSelect = { range = it })

        Row(
            modifier = Modifier.fillMaxWidth(),
            horizontalArrangement = Arrangement.spacedBy(10.dp),
        ) {
            StatCell("Requests", "${filtered.size}", Modifier.weight(1f))
            StatCell("Input", formatTokens(totalInput), Modifier.weight(1f))
            StatCell("Output", formatTokens(totalOutput), Modifier.weight(1f))
        }

        if (filtered.isEmpty()) {
            Text(
                "No usage in this period.",
                color = TextSecondary,
                fontSize = 13.sp,
                textAlign = TextAlign.Center,
                modifier = Modifier.fillMaxWidth(),
            )
        }

        if (byModel.isNotEmpty()) {
            SectionCard(title = "By Model") {
                byModel.forEach { row ->
                    KeyValueRow(
                        title = row.model,
                        subtitle = "${row.requests} request${if (row.requests == 1) "" else "s"}",
                        value = "${formatTokens(row.total)} tokens",
                    )
                }
            }
        }

        if (byApp.isNotEmpty()) {
            SectionCard(title = "By App") {
                byApp.forEach { row ->
                    KeyValueRow(
                        title = hostnameOf(row.origin),
                        subtitle = "${row.requests} request${if (row.requests == 1) "" else "s"}",
                        value = "${formatTokens(row.total)} tokens",
                    )
                }
            }
        }

        SectionCard(title = "Gifts from this credential") {
            if (gifts.isEmpty()) {
                Text(
                    "No gifts created from this credential.",
                    fontSize = 12.sp,
                    color = TextSecondary,
                    modifier = Modifier.padding(vertical = 4.dp),
                )
            } else {
                val totalRedeemed = gifts.sumOf { it.usedTokens }
                Text(
                    "${formatTokens(totalRedeemed)} tokens redeemed across ${gifts.size} gift${if (gifts.size == 1) "" else "s"}",
                    fontSize = 12.sp,
                    color = TextSecondary,
                    modifier = Modifier.padding(bottom = 6.dp),
                )
                gifts.forEach { g ->
                    GiftSummaryRow(gift = g)
                }
            }
        }
    }
}

@Composable
private fun GiftStatsContent(
    credential: GiftedCredential,
    online: Boolean?,
) {
    val provider = Provider.find(credential.providerId)
    val remaining = giftBudgetRemaining(credential.usedTokens, credential.maxTokens)
    val percent = giftBudgetPercent(credential.usedTokens, credential.maxTokens)
    val statusText = when (online) {
        true -> "Online"
        false -> "Offline"
        null -> "Checking…"
    }
    val statusColor = when (online) {
        true -> Color(0xFF34D399)
        false -> Color(0xFFF43F5E)
        null -> Color(0xFFF59E0B)
    }

    Column(
        modifier = Modifier
            .fillMaxWidth()
            .verticalScroll(rememberScrollState())
            .padding(horizontal = 24.dp)
            .padding(bottom = 32.dp),
        verticalArrangement = Arrangement.spacedBy(14.dp),
    ) {
        Column {
            Text(
                "Gift from ${credential.senderLabel}",
                fontSize = 20.sp,
                fontWeight = FontWeight.Bold,
                color = TextPrimary,
            )
            Text(
                provider?.name ?: credential.providerId,
                fontSize = 12.sp,
                color = TextSecondary,
            )
        }

        Row(
            modifier = Modifier.fillMaxWidth(),
            horizontalArrangement = Arrangement.spacedBy(10.dp),
        ) {
            StatCell("Used", formatTokens(credential.usedTokens), Modifier.weight(1f))
            StatCell("Remaining", formatTokens(remaining), Modifier.weight(1f))
            StatCell("Budget", formatTokens(credential.maxTokens), Modifier.weight(1f))
        }

        LinearProgressIndicator(
            progress = { percent / 100f },
            modifier = Modifier
                .fillMaxWidth()
                .height(6.dp)
                .clip(RoundedCornerShape(3.dp)),
            color = if (percent > 80) Warning else Accent,
            trackColor = BgHover,
        )

        SectionCard(title = null) {
            KeyValueRow(title = "Sender", subtitle = null, value = statusText, valueColor = statusColor)
            HorizontalDivider(color = BgHover)
            KeyValueRow(title = "Expires", subtitle = null, value = giftExpiryText(credential.expiresAt))
            HorizontalDivider(color = BgHover)
            KeyValueRow(title = "Received", subtitle = null, value = formatAbsoluteDate(credential.createdAt))
        }
    }
}

@Composable
private fun RangePicker(selected: StatsRange, onSelect: (StatsRange) -> Unit) {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(10.dp))
            .background(BgCard),
    ) {
        StatsRange.entries.forEach { r ->
            val isSelected = r == selected
            Surface(
                onClick = { onSelect(r) },
                modifier = Modifier.weight(1f),
                color = if (isSelected) Accent else BgCard,
                shape = RoundedCornerShape(10.dp),
            ) {
                Text(
                    r.label,
                    modifier = Modifier.padding(vertical = 8.dp),
                    textAlign = TextAlign.Center,
                    fontWeight = FontWeight.Medium,
                    fontSize = 14.sp,
                    color = if (isSelected) TextPrimary else TextSecondary,
                )
            }
        }
    }
}

@Composable
private fun StatCell(label: String, value: String, modifier: Modifier = Modifier) {
    Card(
        modifier = modifier,
        colors = CardDefaults.cardColors(containerColor = BgCard),
        shape = RoundedCornerShape(12.dp),
    ) {
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .padding(vertical = 14.dp),
            horizontalAlignment = Alignment.CenterHorizontally,
        ) {
            Text(value, fontWeight = FontWeight.Bold, fontSize = 18.sp, color = TextPrimary)
            Spacer(Modifier.height(4.dp))
            Text(label, color = TextSecondary, fontSize = 11.sp)
        }
    }
}

@Composable
private fun SectionCard(title: String?, content: @Composable ColumnScope.() -> Unit) {
    Card(
        colors = CardDefaults.cardColors(containerColor = BgCard),
        shape = RoundedCornerShape(12.dp),
    ) {
        Column(modifier = Modifier.padding(14.dp)) {
            if (title != null) {
                Text(title, fontWeight = FontWeight.SemiBold, color = TextPrimary)
                Spacer(Modifier.height(8.dp))
            }
            content()
        }
    }
}

@Composable
private fun KeyValueRow(
    title: String,
    subtitle: String?,
    value: String,
    valueColor: Color = TextPrimary,
) {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .padding(vertical = 6.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Column(modifier = Modifier.weight(1f)) {
            Text(
                title,
                fontSize = 13.sp,
                color = TextPrimary,
                maxLines = 1,
                overflow = TextOverflow.Ellipsis,
            )
            if (subtitle != null) {
                Text(subtitle, fontSize = 11.sp, color = TextSecondary)
            }
        }
        Text(value, fontSize = 13.sp, fontWeight = FontWeight.Medium, color = valueColor)
    }
}

@Composable
private fun GiftSummaryRow(gift: Gift) {
    val percent = giftBudgetPercent(gift.usedTokens, gift.maxTokens)
    val active = gift.active && System.currentTimeMillis() < gift.expiresAt
    Column(modifier = Modifier.padding(vertical = 6.dp)) {
        Row(
            modifier = Modifier.fillMaxWidth(),
            horizontalArrangement = Arrangement.SpaceBetween,
            verticalAlignment = Alignment.CenterVertically,
        ) {
            Text(
                gift.label.ifEmpty { "Unnamed gift" },
                fontSize = 13.sp,
                fontWeight = FontWeight.Medium,
                color = TextPrimary,
            )
            Text(
                if (!active) "Inactive" else giftExpiryText(gift.expiresAt),
                fontSize = 11.sp,
                color = TextSecondary,
            )
        }
        Spacer(Modifier.height(4.dp))
        LinearProgressIndicator(
            progress = { percent / 100f },
            modifier = Modifier
                .fillMaxWidth()
                .height(4.dp)
                .clip(RoundedCornerShape(2.dp)),
            color = if (percent > 80) Warning else Accent,
            trackColor = BgHover,
        )
        Spacer(Modifier.height(4.dp))
        Row(
            modifier = Modifier.fillMaxWidth(),
            horizontalArrangement = Arrangement.SpaceBetween,
        ) {
            Text("${formatTokens(gift.usedTokens)} used", fontSize = 11.sp, color = TextSecondary)
            Text("/ ${formatTokens(gift.maxTokens)}", fontSize = 11.sp, color = TextSecondary)
        }
    }
}

private fun giftExpiryText(expiresAt: Long): String {
    val remaining = expiresAt - System.currentTimeMillis()
    if (remaining <= 0) return "Expired"
    val hours = remaining / 3_600_000
    val days = hours / 24
    return when {
        days > 0 -> "${days}d left"
        hours > 0 -> "${hours}h left"
        else -> "${remaining / 60_000}m left"
    }
}

private fun formatAbsoluteDate(ms: Long): String {
    val fmt = java.text.SimpleDateFormat("d MMM yyyy", java.util.Locale.getDefault())
    return fmt.format(java.util.Date(ms))
}

private fun hostnameOf(origin: String): String =
    try {
        java.net.URI(origin).host ?: origin
    } catch (_: Exception) {
        origin
    }

File: packages/ios/Byoky/Views/Wallet/WalletStatsSheet.swift
import SwiftUI

enum WalletStatsRange: String, CaseIterable {
    case day = "24h"
    case week = "7d"
    case month = "30d"
    case all = "All"

    var interval: TimeInterval? {
        switch self {
        case .day: return 86400
        case .week: return 604800
        case .month: return 2592000
        case .all: return nil
        }
    }
}

struct WalletStatsSheet: View {
    @EnvironmentObject var wallet: WalletStore
    @Environment(\.dismiss) private var dismiss
    let target: WalletStatsTarget

    var body: some View {
        SwiftUI.Group {
            switch target {
            case .credential(let id):
                if let cred = wallet.credentials.first(where: { $0.id == id }) {
                    CredentialStatsView(credential: cred)
                } else {
                    missing
                }
            case .gift(let id):
                if let gc = wallet.giftedCredentials.first(where: { $0.id == id }) {
                    GiftStatsView(credential: gc)
                } else {
                    missing
                }
            }
        }
        .background(Theme.bgMain)
        .toolbarBackground(Theme.bgMain, for: .navigationBar)
        .toolbar {
            ToolbarItem(placement: .cancellationAction) {
                Button("Close") { dismiss() }
            }
        }
    }

    private var missing: some View {
        VStack {
            Text("No data").font(.headline)
        }
        .frame(maxWidth: .infinity, maxHeight: .infinity)
    }
}

private struct CredentialStatsView: View {
    @EnvironmentObject var wallet: WalletStore
    let credential: Credential
    @State private var range: WalletStatsRange = .week

    private var provider: Provider? { Provider.find(credential.providerId) }

    private var logs: [RequestLog] {
        let all = wallet.requestLogs.filter { $0.providerId == credential.providerId && $0.statusCode < 400 }
        guard let interval = range.interval else { return all }
        let cutoff = Date().addingTimeInterval(-interval)
        return all.filter { $0.timestamp > cutoff }
    }

    private var totalInput: Int { logs.reduce(0) { $0 + ($1.inputTokens ?? 0) } }
    private var totalOutput: Int { logs.reduce(0) { $0 + ($1.outputTokens ?? 0) } }

    private var byModel: [(model: String, total: Int, requests: Int)] {
        var map: [String: (total: Int, requests: Int)] = [:]
        for e in logs {
            guard let m = e.model else { continue }
            let prev = map[m] ?? (0, 0)
            map[m] = (prev.total + (e.inputTokens ?? 0) + (e.outputTokens ?? 0), prev.requests + 1)
        }
        return map.map { (model: $0.key, total: $0.value.total, requests: $0.value.requests) }
            .sorted { $0.total > $1.total }
    }

    private var byApp: [(origin: String, total: Int, requests: Int)] {
        var map: [String: (total: Int, requests: Int)] = [:]
        for e in logs {
            let prev = map[e.appOrigin] ?? (0, 0)
            map[e.appOrigin] = (prev.total + (e.inputTokens ?? 0) + (e.outputTokens ?? 0), prev.requests + 1)
        }
        return map.map { (origin: $0.key, total: $0.value.total, requests: $0.value.requests) }
            .sorted { $0.total > $1.total }
    }

    private var gifts: [Gift] {
        wallet.gifts.filter { $0.credentialId == credential.id }
    }

    private var giftUsedTotal: Int { gifts.reduce(0) { $0 + $1.usedTokens } }

    var body: some View {
        ScrollView {
            VStack(alignment: .leading, spacing: 16) {
                header

                rangePicker

                totalsRow

                if logs.isEmpty {
                    emptySection
                }

                if !byModel.isEmpty {
                    section("By Model") {
                        ForEach(byModel, id: \.model) { row in
                            StatRow(
                                title: row.model,
                                subtitle: "\(row.requests) request\(row.requests == 1 ? "" : "s")",
                                trailing: formatWalletTokens(row.total) + " tokens",
                                monospace: true
                            )
                        }
                    }
                }

                if !byApp.isEmpty {
                    section("By App") {
                        ForEach(byApp, id: \.origin) { row in
                            StatRow(
                                title: hostname(row.origin),
                                subtitle: "\(row.requests) request\(row.requests == 1 ? "" : "s")",
                                trailing: formatWalletTokens(row.total) + " tokens"
                            )
                        }
                    }
                }

                section("Gifts from this credential") {
                    if gifts.isEmpty {
                        Text("No gifts created from this credential.")
                            .font(.caption)
                            .foregroundStyle(.secondary)
                            .padding(.vertical, 4)
                    } else {
                        Text("\(formatWalletTokens(giftUsedTotal)) tokens redeemed across \(gifts.count) gift\(gifts.count == 1 ? "" : "s")")
                            .font(.caption)
                            .foregroundStyle(.secondary)
                            .padding(.bottom, 4)
                        ForEach(gifts) { g in
                            GiftSummaryRow(gift: g)
                        }
                    }
                }
            }
            .padding()
        }
        .navigationTitle(credential.label)
        .navigationBarTitleDisplayMode(.inline)
    }

    private var header: some View {
        VStack(alignment: .leading, spacing: 4) {
            Text(provider?.name ?? credential.providerId)
                .font(.subheadline.weight(.semibold))
            Text("Shared across all credentials of this provider")
                .font(.caption)
                .foregroundStyle(.secondary)
        }
    }

    private var rangePicker: some View {
        HStack(spacing: 0) {
            ForEach(WalletStatsRange.allCases, id: \.self) { r in
                Button {
                    withAnimation(.easeInOut(duration: 0.2)) { range = r }
                } label: {
                    Text(r.rawValue)
                        .font(.subheadline.weight(.medium))
                        .frame(maxWidth: .infinity)
                        .padding(.vertical, 8)
                        .background(range == r ? Theme.accent : Color.clear)
                        .foregroundStyle(range == r ? .white : .secondary)
                }
            }
        }
        .clipShape(RoundedRectangle(cornerRadius: 10))
        .background(Theme.bgRaised)
        .overlay(RoundedRectangle(cornerRadius: 10).stroke(Theme.border, lineWidth: 1))
        .clipShape(RoundedRectangle(cornerRadius: 10))
    }

    private var totalsRow: some View {
        HStack(spacing: 12) {
            StatCard(value: "\(logs.count)", label: "Requests")
            StatCard(value: formatWalletTokens(totalInput), label: "Input")
            StatCard(value: formatWalletTokens(totalOutput), label: "Output")
        }
    }

    private var emptySection: some View {
        Text("No usage in this period.")
            .font(.callout)
            .foregroundStyle(.secondary)
            .frame(maxWidth: .infinity, alignment: .center)
            .padding(.vertical, 8)
    }

    private func section<Content: View>(_ title: String, @ViewBuilder content: () -> Content) -> some View {
        VStack(alignment: .leading, spacing: 8) {
            Text(title)
                .font(.headline)
            content()
        }
        .padding(14)
        .background(Theme.bgCard)
        .overlay(RoundedRectangle(cornerRadius: 12).stroke(Theme.border, lineWidth: 1))
        .clipShape(RoundedRectangle(cornerRadius: 12))
    }

    private func hostname(_ origin: String) -> String {
        URL(string: origin)?.host ?? origin
    }
}

private struct GiftStatsView: View {
    @EnvironmentObject var wallet: WalletStore
    let credential: GiftedCredential

    private var provider: Provider? { Provider.find(credential.providerId) }
    private var remaining: Int { giftedBudgetRemaining(credential) }
    private var percent: Double { giftedBudgetPercent(credential) }
    private var onlineState: Bool? { wallet.giftPeerOnline[credential.giftId] }

    private var onlineText: String {
        switch onlineState {
        case .some(true): return "Online"
        case .some(false): return "Offline"
        case nil: return "Checking…"
        }
    }

    private var onlineColor: Color {
        switch onlineState {
        case .some(true): return .green
        case .some(false): return .red
        case nil: return .orange
        }
    }

    private var expiryText: String {
        if isGiftedCredentialExpired(credential) { return "Expired" }
        let formatter = RelativeDateTimeFormatter()
        formatter.unitsStyle = .abbreviated
        return formatter.localizedString(for: credential.expiresAt, relativeTo: Date())
    }

    var body: some View {
        ScrollView {
            VStack(alignment: .leading, spacing: 16) {
                VStack(alignment: .leading, spacing: 4) {
                    Text(provider?.name ?? credential.providerId)
                        .font(.subheadline.weight(.semibold))
                    Text("From \(credential.senderLabel)")
                        .font(.caption)
                        .foregroundStyle(.secondary)
                }

                HStack(spacing: 12) {
                    StatCard(value: formatWalletTokens(credential.usedTokens), label: "Used")
                    StatCard(value: formatWalletTokens(remaining), label: "Remaining")
                    StatCard(value: formatWalletTokens(credential.maxTokens), label: "Budget")
                }

                ProgressView(value: percent)
                    .tint(percent > 0.9 ? .red : Theme.accent)

                VStack(spacing: 0) {
                    StatRow(title: "Sender",
                            subtitle: nil,
                            trailing: onlineText,
                            trailingColor: onlineColor)
                    Divider()
                    StatRow(title: "Expires",
                            subtitle: nil,
                            trailing: expiryText)
                    Divider()
                    StatRow(title: "Received",
                            subtitle: nil,
                            trailing: credential.createdAt.formatted(date: .abbreviated, time: .omitted))
                }
                .padding(14)
                .background(Theme.bgCard)
                .overlay(RoundedRectangle(cornerRadius: 12).stroke(Theme.border, lineWidth: 1))
                .clipShape(RoundedRectangle(cornerRadius: 12))
            }
            .padding()
        }
        .navigationTitle("Gift")
        .navigationBarTitleDisplayMode(.inline)
    }
}

private struct StatCard: View {
    let value: String
    let label: String

    var body: some View {
        VStack(spacing: 4) {
            Text(value)
                .font(.title3.weight(.bold).monospacedDigit())
            Text(label)
                .font(.caption)
                .foregroundStyle(.secondary)
        }
        .frame(maxWidth: .infinity)
        .padding(.vertical, 14)
        .background(Theme.bgCard)
        .overlay(RoundedRectangle(cornerRadius: 12).stroke(Theme.border, lineWidth: 1))
        .clipShape(RoundedRectangle(cornerRadius: 12))
    }
}

private struct StatRow: View {
    let title: String
    var subtitle: String? = nil
    let trailing: String
    var trailingColor: Color? = nil
    var monospace: Bool = false

    var body: some View {
        HStack(alignment: .firstTextBaseline) {
            VStack(alignment: .leading, spacing: 2) {
                Text(title)
                    .font(monospace ? .caption.monospaced() : .callout)
                    .lineLimit(1)
                    .truncationMode(.middle)
                if let subtitle {
                    Text(subtitle)
                        .font(.caption2)
                        .foregroundStyle(.secondary)
                }
            }
            Spacer()
            Text(trailing)
                .font(.callout.monospacedDigit())
                .foregroundStyle(trailingColor ?? .primary)
        }
        .padding(.vertical, 6)
    }
}

private struct GiftSummaryRow: View {
    let gift: Gift

    private var percent: Double {
        guard gift.maxTokens > 0 else { return 0 }
        return min(1, Double(gift.usedTokens) / Double(gift.maxTokens))
    }

    private var expiryText: String {
        if !gift.active || Date() > gift.expiresAt { return "Inactive" }
        let formatter = RelativeDateTimeFormatter()
        formatter.unitsStyle = .abbreviated
        return "Expires " + formatter.localizedString(for: gift.expiresAt, relativeTo: Date())
    }

    var body: some View {
        VStack(alignment: .leading, spacing: 6) {
            HStack {
                Text(gift.label.isEmpty ? "Unnamed gift" : gift.label)
                    .font(.callout.weight(.medium))
                Spacer()
                Text(expiryText)
                    .font(.caption2)
                    .foregroundStyle(.secondary)
            }
            ProgressView(value: percent)
                .tint(percent > 0.9 ? .red : Theme.accent)
            HStack {
                Text("\(formatWalletTokens(gift.usedTokens)) used")
                    .font(.caption2)
                    .foregroundStyle(.secondary)
                Spacer()
                Text("/ \(formatWalletTokens(gift.maxTokens))")
                    .font(.caption2)
                    .foregroundStyle(.secondary)
            }
        }
        .padding(.vertical, 6)
    }
}
